include-blang/bugle.h

```diff
@@ -12,6 +12,7 @@ void _DEVICE_QUALIFIER bugle_assume(int);
 void _DEVICE_QUALIFIER bugle_requires(int);
 void _DEVICE_QUALIFIER bugle_ensures(int);
 void _DEVICE_QUALIFIER bugle_barrier(bool local_flags, bool global_flags);
+void _DEVICE_QUALIFIER bugle_grid_barrier();
```

Comment:
Alternatively to introducing a new function, we could add an extra flag to `bugle_barrier`, but this seems reasonable.

Reply:
Having looked at bit at https://github.com/mc-imperial/gpuverify/pull/40, I do think things there would be easier if this function is given the same arguments as `bugle_barrier`.

Reply:
Can you please elaborate on this? How would we use the local_flags and global_flags arguments in bugle_grid_barrier? For the CUDA implementation of bugle_barrier, we are setting them as true in Bugle since the concept of local and global fences is exclusive to OpenCL.

Reply:
Indeed we're setting them `true` always. My main aim is to reduce the amount of special casing that needs to happen in GPUVerify. Ideally, there we go to a situation where we can leverage everything that is in place for `bugle_barrier`, but with one small change, namely that the behavior for threads in different blocks changes (this should be just one or two small changes to the generated conditions).

Reply:
I still want this change, but I'll make this change myself once this PR is merged.

Reply:
I am working on this. This will also need changes in GPUVerify. I will make all the changes and run the tests before committing them. I didn't mark this comment as resolved because of that. It might take me a little while since the changes are widespread.

Reply:
If I'm not mistaken you should be able to make this change without breaking your changes on the GPUVerify side. Those could be cleaned up though, but we can do that separately.

Reply:
Just to confirm, are you suggesting 1) using bugle_barrier with a third argument or 2) using bugle_grid_barrier with two arguments (both set to true in bugle)? I am implementing the first option, that will need some changes in GPUVerify.

Reply:
I'm suggesting:
```
void _DEVICE_QUALIFIER bugle_grid_barrier(bool local_flags, bool global_flags);
```

Reply:
And then every call becomes:
```
bugle_grid_barrier(true, true);
```

Reply:
Got it, Thank you very much for the confirmation. I will implement this and commit the changes.

Reply:
This change broke the PR in GPUVerify. I will fix that tomorrow and send an update.